### PR TITLE
[vs_platform_toolset msbuild] MSVC ClangCL [Only TEST, no merge]

### PR DESCRIPTION
**Pull request by MVoz**
Only TEST, no merge

test only triplet x64-windows, only **msbuild project** and cmake -T"" **msbuild** // CMAKE_GENERATOR "Visual Studio"

### CODE REVIEW

**Comment by Neumann-A** [comment]
In general that many changes are not required.

what you can do is use the following triplet:

```
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE dynamic)
if(DEFINED VCPKG_PLATFORM_TOOLSET)
 # this lets vcpkg initialize the normal VS enviroment 
 # before switching to llvm while building the ports
 set(VCPKG_PLATFORM_TOOLSET llvm) # requires llvm extension in vs
else()
 set(VCPKG_PLATFORM_TOOLSET v142) # only required if you don't want to use vcpkg auto detection
endif()
set(VCPKG_POLICY_SKIP_ARCHITECTURE_CHECK enabled) # required to skip missing 2.linker member error
# For ninja and other ports:
set(ENV{CC} "<path/to/clang-cl>")
set(ENV{CXX} "<path/to/clang-cl>")
```

**Comment by Voskrese** [comment]
ClangCL extension is already integrated into MSVC

File: scripts/cmake/vcpkg_install_msbuild.cmake [hunk 1]
```diff
@@ -132,21 +135,55 @@ function(vcpkg_install_msbuild)
...
+        "/p:BasePlatformToolset=${VCPKG_PLATFORM_TOOLSET}"
+        "/p:PlatformToolset=${VS_PLATFORM_TOOLSET}"
+        "/p:UseIntelMKL=No"
+        "/p:UseClangCl=true"
+        "/p:UseLldLink=true"
+        "/p:UseLlvmLib=false"
+        "/p:TrackFileAccess=false"
+        "/p:LibToolExe=lib.exe"
+        "/p:ClangClAdditionalOptions=-Wno-gcc-compat -Xclang -fopenmp -Xclang -flto=thin -Wextra -Wno-unused-variable -verbose"
+        "/p:LldLinkAdditionalOptions=-debug -force:multipleres -fuse-ld=lld-link -lib"
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_install_msbuild.cmake:160]: These should probably be VCPKG_ADDITIONAL_MSBUILD_OPTIONS so they get settable from the triplet

File: scripts/cmake/vcpkg_install_msbuild.cmake [hunk 2]
```diff
@@ -132,21 +135,55 @@ function(vcpkg_install_msbuild)
...
         # TODO: Add a way for the user to override this if they want to opt-in to incompatibility
         list(APPEND _csc_OPTIONS /p:WholeProgramOptimization=false)
     endif()
 
     if(_csc_USE_VCPKG_INTEGRATION)
-        list(APPEND _csc_OPTIONS /p:ForceImportBeforeCppTargets=${SCRIPTS}/buildsystems/msbuild/vcpkg.targets /p:VcpkgApplocalDeps=false)
+      list(APPEND _csc_OPTIONS
+        "/p:ForceImportBeforeCppTargets=${SCRIPTS}/buildsystems/msbuild/vcpkg.targets"
+        "/p:VcpkgApplocalDeps=false"
+        "/p:VcpkgTriplet=${TARGET_TRIPLET}"
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_install_msbuild.cmake:185]: +1. That one is truely missing.

**Comment by Voskrese** [comment]
@Neumann-A 
how do I merge extra commits ?)))

**Comment by Neumann-A** [comment]
I normally do 
```
git remote add <somename> <someupstreamgit>
git fetch <somename>
git merge <somename>/<somebranch>
```
If the remote is already added the first command can be skipped
So for vcpkg master I normally do:
```
git remote add upstream https://github.com/microsoft/vcpkg.git
git fetch upstream
git merge upstream/master
```
the first command is normally skipped since I already added the upstream remote add some point earlier.

**Comment by Voskrese** [comment]
> if(DEFINED VCPKG_PLATFORM_TOOLSET)
>  # this lets vcpkg initialize the normal VS enviroment 
>  # before switching to llvm while building the ports
>  set(VCPKG_PLATFORM_TOOLSET llvm) # requires llvm extension in vs
> else()
>  set(VCPKG_PLATFORM_TOOLSET v142) # only required if you don't want to use vcpkg auto detection
> endif()

this won't work, MSBUILD needs to have two Platformtoolsets \\ BasePlatformtoolsets

**Comment by Voskrese** [comment]
not install
https://docs.microsoft.com/en-us/cpp/build/clang-support-msbuild?view=vs-2019
https://devblogs.microsoft.com/cppblog/clang-llvm-support-in-visual-studio/

no test (((

**Comment by Neumann-A** [comment]
I use this
https://marketplace.visualstudio.com/items?itemName=LLVMExtensions.llvm-toolchain
or now this for VS 2019
https://marketplace.visualstudio.com/items?itemName=MarekAniola.mangh-llvm2019

>/p:BasePlatformToolset=v141

introduce `VCPKG_ADDITIONAL_MSBUILD_OPTIONS` and you can pass whatever you like via the triplet file. But more importantly would be to have something like `VCPKG_ENVIROMNENT_SETUP_SCRIPT` for the Intel compiler

**Comment by Voskrese** [comment]
extension installed cmd
https://github.com/zufuliu/llvm-utils/blob/master/VS2017/install.bat

**Comment by Voskrese** [comment]
@Neumann-A 
installed extension user folder ?

DISK`:\Users\Nikolay\AppData\Local\Microsoft\MSBuild`

```
├───10.0
│   │   Microsoft.Cpp.UpgradeFromVC60.props
│   │   Microsoft.Cpp.UpgradeFromVC70.props
│   │   Microsoft.Cpp.UpgradeFromVC71.props
│   │   Microsoft.Cpp.UpgradeFromVC80.props
│   │   Microsoft.Cpp.UpgradeFromVC81.props
│   │
│   └───Microsoft.Common.targets
│       ├───ImportAfter
│       └───ImportBefore
│               SonarQube.Integration.ImportBefore.targets
│
├───11.0
│   │   Microsoft.Cpp.UpgradeFromVC60.props
│   │   Microsoft.Cpp.UpgradeFromVC70.props
│   │   Microsoft.Cpp.UpgradeFromVC71.props
│   │   Microsoft.Cpp.UpgradeFromVC80.props
│   │   Microsoft.Cpp.UpgradeFromVC81.props
│   │
│   └───Microsoft.Common.targets
│       ├───ImportAfter
│       └───ImportBefore
│               SonarQube.Integration.ImportBefore.targets
│
├───12.0
│   └───Microsoft.Common.targets
│       ├───ImportAfter
│       └───ImportBefore
│               SonarQube.Integration.ImportBefore.targets
│
├───14.0
│   └───Microsoft.Common.targets
│       ├───ImportAfter
│       └───ImportBefore
│               SonarQube.Integration.ImportBefore.targets
│
├───15.0
│   └───Microsoft.Common.targets
│       └───ImportBefore
│               SonarQube.Integration.ImportBefore.targets
│
├───4.0
│   │   Microsoft.Cpp.UpgradeFromVC60.props
│   │   Microsoft.Cpp.UpgradeFromVC70.props
│   │   Microsoft.Cpp.UpgradeFromVC71.props
│   │   Microsoft.Cpp.UpgradeFromVC80.props
│   │   Microsoft.Cpp.UpgradeFromVC81.props
│   │
│   └───Microsoft.Common.targets
│       ├───ImportAfter
│       └───ImportBefore
│               SonarQube.Integration.ImportBefore.targets
│
├───9.0
│   │   Microsoft.Cpp.UpgradeFromVC60.props
│   │   Microsoft.Cpp.UpgradeFromVC70.props
│   │   Microsoft.Cpp.UpgradeFromVC71.props
│   │   Microsoft.Cpp.UpgradeFromVC80.props
│   │   Microsoft.Cpp.UpgradeFromVC81.props
│   │
│   └───Microsoft.Common.targets
│       ├───ImportAfter
│       └───ImportBefore
│               SonarQube.Integration.ImportBefore.targets
│
└───v4.0
        Microsoft.Cpp.Android.user.props
        Microsoft.Cpp.ARM.user.props
        Microsoft.Cpp.ARM64.user.props
        Microsoft.Cpp.chpe.user.props
        Microsoft.Cpp.Clang.user.props
        Microsoft.Cpp.Emscripten.user.props
        Microsoft.Cpp.iOS.user.props
        Microsoft.Cpp.Itanium.user.props
        Microsoft.Cpp.MinGW.user.props
        Microsoft.Cpp.MinGW32.user.props
        Microsoft.Cpp.MinGW64.user.props
        Microsoft.Cpp.NaCl.user.props
        Microsoft.Cpp.NaCl32.user.props
        Microsoft.Cpp.NaCl64.user.props
        Microsoft.Cpp.NaClARM.user.props
        Microsoft.Cpp.PNaCl.user.props
        Microsoft.Cpp.PPAPI.user.props
        Microsoft.Cpp.pythonemcc.user.props
        Microsoft.Cpp.tmp.user.props
        Microsoft.Cpp.Win32.user.props
        Microsoft.Cpp.x64.user.props
```

**Comment by Voskrese** [comment]
@BillyONeal 
The build tools for ClangCL (Platform Toolset = 'ClangCL') cannot be found. To build using the ClangCL build tools, please install ClangCL build tools.

not installed in Ci
C++ Clang tools for Windows ,  C++ Clang-cl for v142 build tools
https://docs.microsoft.com/en-us/cpp/build/clang-support-msbuild?view=vs-2019

close PR

**Pull request closed without merging**